<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [6]:
# текст запроса
# Просчитываем общее кол.-во вакансий по столбецу "id", данные заносим в созданый столбец "Кол-во вакансий"
query_3_1 = f'''SELECT COUNT(id) AS "Кол-во вакансий" 
                FROM public.vacancies''' # Из схеме public таблицы "vacancies"
# Используем f-строки для удобства передачи параметров

In [7]:
# результат запроса
# Используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df_vacancies = pd.read_sql_query(query_3_1, connection)
df_vacancies

C:\Users\Professional\AppData\Local\Temp\ipykernel_9196\285274537.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_vacancies = pd.read_sql_query(query_3_1, connection)


Кол-во вакансий
0            49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [8]:
# текст запроса
# Просчитываем общее кол.-во работодателей по столбецу "name", данные заносим в созданый столбец "Кол-во работодателей"
query_3_2 = f'''SELECT COUNT(name) AS "Кол-во работодателей"
                FROM public.employers''' # Из схеме public таблицы "employers"
# Используем f-строки для удобства передачи параметров

In [9]:
# результат запроса
# Используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df_employers = pd.read_sql_query(query_3_2, connection)
df_employers

C:\Users\Professional\AppData\Local\Temp\ipykernel_9196\144432154.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_employers = pd.read_sql_query(query_3_2, connection)


Кол-во работодателей
0                 23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [10]:
# текст запроса
# Просчитываем общее кол.-во регионов используемых для анализа по столбецу "id", данные заносим в созданый столбец "Кол-во регионов"
query_3_3 = f'''SELECT COUNT(id) AS "Кол-во регионов"
                FROM public.areas''' # Из схеме public таблицы "areas"
# Используем f-строки для удобства передачи параметров

In [11]:
# результат запроса
# Используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df_areas = pd.read_sql_query(query_3_3, connection)
df_areas

C:\Users\Professional\AppData\Local\Temp\ipykernel_9196\605157076.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_areas = pd.read_sql_query(query_3_3, connection)


Кол-во регионов
0             1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [12]:
# текст запроса
# Просчитываем общее кол.-во сфер деятельности попавших в выборку по столбецу "name", данные заносим в созданый столбец "Кол-во сфер деятельности"
query_3_4 = f'''SELECT 
                    COUNT(name) AS "Кол-во сфер деятельности"
                FROM public.industries''' # Из схеме public таблицы "industries"
# Используем f-строки для удобства передачи параметров

In [13]:
# результат запроса
# Используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df_industries = pd.read_sql_query(query_3_4, connection)
df_industries

C:\Users\Professional\AppData\Local\Temp\ipykernel_9196\1910758124.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_industries = pd.read_sql_query(query_3_4, connection)


Кол-во сфер деятельности
0                       294

***

# выводы по предварительному анализу данных
По итогам предворительного анализа можно сказать:
 - Присудствие достаточно большого количества вакансий (49197) в базе данных может указывать на активность рынка труда и наличие разнообразных возможностей для соискателей.
 - Большое количество компаний и организаций (23501) представленых в базе данных как работодалели, а так же их географическое расположение(база данных содержит информацию о 1362 регионах) указывает на возможность выбора соикателем как в плане выбора работодателя, так и в плане выбора самого места работы.
 - Сферы деятельности представлены в базе данных 294. Это свидетельствует о разнообразии отраслей и секторов экономики которые нуждаються в рабочих кадрах.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [16]:
# текст запроса
query_4_1 = f'''SELECT
                    COUNT(v.id) AS cnt, /* Просчитываем общее кол.-во вакансий по столбецу "id" из таблици v, данные заносим в созданый столбец "cnt"*/
                    a.name AS area /* Указываем регион по столбецу "name" из таблици a, данные заносим в созданый столбец "area"*/
                FROM public.vacancies v /*Из схеме public таблицы "vacancies", задаём таблице алиас v*/
                JOIN public.areas a ON a.id = v.area_id /*Внутреннее присоединение из схемы  public таблицы areas, задав ей алиас a, по ключам  id и area_id*/
                GROUP BY 2 /* Групируем п оназванию региона*/
                ORDER BY 1 DESC /*Сортируем по количеству вакансий в порядке убывания*/
                LIMIT 5''' # Выводим первые 5 регионов с максимальным количеством вакансий
# Используем f-строки для удобства передачи параметров

In [17]:
# результат запроса
# Используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df_cnt= pd.read_sql_query(query_4_1, connection)
df_cnt

C:\Users\Professional\AppData\Local\Temp\ipykernel_9196\2631575504.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_cnt= pd.read_sql_query(query_4_1, connection)


cnt             area
0  5333           Москва
1  2851  Санкт-Петербург
2  2112            Минск
3  2006      Новосибирск
4  1892           Алматы

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [18]:
# текст запроса
query_4_2 = f'''SELECT 
                    COUNT(*) AS "Кол-во вакансий" /* Просчитываем кол.-во вакансий, данные заносим в созданый столбец "Кол-во вакансий"*/
                FROM public.vacancies v /*Из схеме public таблицы "vacancies", задаём таблице алиас v*/
                WHERE v.salary_from IS NOT NULL OR v.salary_to IS NOT NULL''' # Указываем услови при котором з/п указано хотя бы в одной из зарплатых вилок
# Используем f-строки для удобства передачи параметров

In [19]:
# результат запроса
# Используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df_cnt_v = pd.read_sql_query(query_4_2, connection)
df_cnt_v

C:\Users\Professional\AppData\Local\Temp\ipykernel_9196\2152375433.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_cnt_v = pd.read_sql_query(query_4_2, connection)


Кол-во вакансий
0            24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [20]:
# текст запроса
query_4_3 = f'''SELECT
                    ROUND(AVG(v.salary_from)) AS "Среднее значение для нижней границы зп вилки", /* Определяем среднее значение нижней границы зарплатной вилки и заносим в соотведствующий столбец*/
                    ROUND(AVG(v.salary_to)) AS "Среднее значение для верхней границы зп вилки" /* Определяем среднее значение верхней границы зарплатной вилки и заносим в соотведствующий столбец*/
                FROM public.vacancies v''' # Из схеме public таблицы "vacancies", задаём таблице алиас v*
# Используем f-строки для удобства передачи параметров

In [21]:
# результат запроса
# Используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df_avg = pd.read_sql_query(query_4_3, connection)
df_avg

C:\Users\Professional\AppData\Local\Temp\ipykernel_9196\572238135.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_avg = pd.read_sql_query(query_4_3, connection)


Среднее значение для нижней грани  Среднее значение для верхней гран
0                            71065.0                           110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [41]:
# текст запроса
query_4_4 = f'''SELECT 
                    COUNT(v.id) AS "Кол-во вакансий", /*Просчитываем общее кол.-во вакансий по столбецу "id" из таблици v,*/
                                                         /*данные заносим в созданый столбец "Кол-во вакансий"*/
                    v.schedule AS "Тип рабочего графика", /*Определяем рабочий график по столбецу "schedule" из таблици v,*/
                                                         /*данные заносим в созданый столбец "Тип рабочего графика"*/
                    v.employment AS "Тип трудоустройства" /*Определяем способ трудоустройства по столбецу "employment" из таблици v,*/
                                                          /*данные заносим в созданый столбец "Тип трудоустройства"*/
                 FROM public.vacancies v /*Из схеме public таблицы "vacancies", задаём таблице алиас v*/
                 GROUP BY v.schedule, v.employment /*Группируем таблицу по второму и третему столбцу*/
                 ORDER BY 1 DESC /*Сортируем по количеству вакансий в порядке убывания*/
                 LIMIT 2''' # Выводим первые две пары График/трудоустройство с максимальным количеством вакансий
# Используем f-строки для удобства передачи параметров

In [42]:
# результат запроса
# Используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\Professional\AppData\Local\Temp\ipykernel_9196\2999877184.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


Кол-во вакансий Тип рабочего графика Тип трудоустройства
0            35367          Полный день    Полная занятость
1             7802     Удаленная работа    Полная занятость

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [45]:
# текст запроса
query_4_5 = f'''SELECT 
                    COUNT(v.id) AS cnt, /*Просчитываем общее кол.-во вакансий по столбецу "id" из таблици v, данные заносим в созданый столбец "cnt"*/
                    v.experience AS "Требуемый опыт работы" /*Определяем запрвшиваемый опыт работы по столбецу "experience" из таблици v,*/
                                                            /*данные заносим в созданый столбец "Требуемый опыт работы"*/
                 FROM public.vacancies v /*Из схеме public таблицы "vacancies", задаём таблице алиас v*/
                 GROUP BY v.experience /*Группируем по опыту работы*/
                 ORDER BY 1''' #Сортируем по количеству вакансий в порядке возростания
# Используем f-строки для удобства передачи параметров

In [46]:
# результат запроса
# Используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df_cnt_exp = pd.read_sql_query(query_4_5, connection)
df_cnt_exp

C:\Users\Professional\AppData\Local\Temp\ipykernel_9196\2080874539.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_cnt_exp = pd.read_sql_query(query_4_5, connection)


cnt Требуемый опыт работы
0   1337           Более 6 лет
1   7197             Нет опыта
2  14511         От 3 до 6 лет
3  26152    От 1 года до 3 лет

# выводы по детальному анализу вакансий
После боллее подробного анализа предоставленных даных можно сказать следующее
 - Пятёркой лидеров по поиску сотрудников являються Москва, Санкт-Петербург, Минск, Новосибирск и Алматы. Эти города, в которых предложение вакансий наиболее высоко, что может указывать на их значимость и развитость как центров бизнеса и экономики.
 - Информация о зарплате указана для 24073 вакансий. Это позволяет провести анализ зарплатного уровня и оценить рыночную стоимость работы.
 - Среднее значение для нижней зарплатной вилки составляет 71065, а для верхней зарплатной вилки - 110537. Это диапазон зарплат может служить ориентиром для соискателей в планировании своей заработной платы.
 - Опыт работы от 1 года до 3 лет является наиболее востребованным. Это может указывать на то, что многие работодатели ищут кандидатов с определенным уровнем опыта, который считается оптимальным для выполнения задач в соответствующей области.
 - Полная занятость, как наиболее популярный тип занятости, указывает на то что большинство работодателей ищут сотрудников, готовых работать полный рабочий день.

***

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [62]:
# текст запроса
query_5_1 = f'''(SELECT 
                     e.name AS "Работодатель", /*Выбираем работодателя по столбцу "name" из таблици e, данные заносим в созданый столбец "Работодатель"*/
                     COUNT(v.id) AS "Кол-во вакансий" /*Просчитываем общее кол.-во вакансий по столбецу "id" из таблици v, данные заносим в созданый столбец "Кол-во вакансий"*/
                 FROM public.employers e /*Из схеме public таблицы "employers", задаём таблице алиас e*/
                 JOIN public.vacancies v ON v.employer_id = e.id /*Внутреннее присоединение из схемы public таблицы vacancies, задав ей алиас v, по ключам id и employer_id*/
                 GROUP BY e.name /*Группируем по столбцу "Работодатель"*/
                 ORDER BY 2 DESC /*Сортируем по количеству вакансий в порядке убывания*/
                 LIMIT 1)  /*Выводим первую строку данной выборки*/
                 UNION ALL /*Оператор присоединения*/
                 (SELECT 
                     e.name AS "Работодатель", /*Выбираем работодателя по столбцу "name" из таблици e, данные заносим в созданый столбец "Работодатель"*/
                     COUNT(v.id) AS "Кол-во вакансий" /*Просчитываем общее кол.-во вакансий по столбецу "id" из таблици v, данные заносим в созданый столбец "Кол-во вакансий"*/
                 FROM public.employers e /*Из схеме public таблицы "employers", задаём таблице алиас e*/
                 JOIN public.vacancies v ON v.employer_id = e.id /*Внутреннее присоединение из схемы public таблицы vacancies, задав ей алиас v, по ключам id и employer_id*/
                 GROUP BY e.name /*Группируем по столбцу "Работодатель"*/
                 ORDER BY 2 DESC /*Сортируем по количеству вакансий в порядке убывания*/
                 LIMIT 1 /*Выводим первую строку данной выборки*/
                 OFFSET 4) /*Указываем отчет с пятой строки*/'''
# Используем f-строки для удобства передачи параметров

In [63]:
# результат запроса
# Используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df_emp = pd.read_sql_query(query_5_1, connection)
df_emp

C:\Users\Professional\AppData\Local\Temp\ipykernel_9196\29496228.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_emp = pd.read_sql_query(query_5_1, connection)


Работодатель  Кол-во вакансий
0         Яндекс             1933
1  Газпром нефть              331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [69]:
# текст запроса
query_5_2 = f'''SELECT 
                     a.name AS "Регион", /*Выбираем регион по столбцу "name" из таблици a, данные заносим в созданый столбец "Регион"*/
                     COUNT(e.name) AS "Кол-во работодателей", /*Просчитываем кол.-во работодателей по столбецу "name" из таблици e,*/
                                                              /*данные заносим в созданый столбец "Кол-во работодателей"*/
                     COUNT(v.name) AS "Кол-во вакансий" /*Просчитываем кол.-во вакансий по столбецу "name" из таблици v,*/
                                                        /*данные заносим в созданый столбец "Кол-во вакансий"*/
                 FROM public.areas a /*Из схеме public таблицы "areas", задаём таблице алиас a*/
                 JOIN public.employers e ON a.id = e.area /*Внутреннее присоединение из схемы public таблицы employers,*/
                                                          /*задав ей алиас e, по ключам id и area*/
                 LEFT JOIN public.vacancies v ON v.area_id = a.id /*Внутреннее присоединение из схемы public таблицы vacancies,*/
                                                                 /*задав ей алиас v, по ключам id и area_id*/
                 WHERE v.area_id IS NULL /*Указываем услови при котором вакансии в регионе нет*/
                 GROUP BY 1 /*Группируем по столбецу "Регион"*/
                 ORDER BY 2 DESC /*Сотрируем по столбецу "Кол-во работодателей" в порядке убывания*/
                 LIMIT 3 /*Выводим первые три строки запроса*/''' 
# Используем f-строки для удобства передачи параметров

In [70]:
# результат запроса
# Используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df_регион= pd.read_sql_query(query_5_2, connection)
df_регион

C:\Users\Professional\AppData\Local\Temp\ipykernel_9196\1210127603.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_регион= pd.read_sql_query(query_5_2, connection)


Регион  Кол-во работодателей  Кол-во вакансий
0              Россия                   410                0
1           Казахстан                   207                0
2  Московская область                    75                0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [71]:
# текст запроса
query_5_3 = f'''SELECT       
                     e.name AS "Работодатель", /*Выбираем работодателей по столбецу "name" из таблици e,*/
                                               /*данные заносим в созданый столбец "Работодатель"*/      
                     COUNT(DISTINCT v.area_id) AS "Кол-во регионов" /*Подсчитываем кол.-во регионов в котором работодатель опубликовал вакансии,*/
                                              /*данные заносим в созданый столбец "Кол-во регионов"*/ 
                FROM public.vacancies v /*Из схеме public таблицы "vacancies", задаём таблице алиас v*/
                FULL JOIN public.employers e ON v.employer_id = e.id /*Внутреннее присоединение из схемы public таблицы employers,*/
                                                          /*задав ей алиас e, по ключам id и employer_id*/
                GROUP BY 1 /*Группирум по столбецу "Работодатель"*/
                ORDER BY 2 DESC /*Сортируем по столбцу "Кол-во регионов" в порядке убывания*/ ''' 
# Используем f-строки для удобства передачи параметров

In [72]:
# результат запроса
# Используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df_5_3 = pd.read_sql_query(query_5_3, connection)
df_5_3

C:\Users\Professional\AppData\Local\Temp\ipykernel_9196\2173856228.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_3 = pd.read_sql_query(query_5_3, connection)


Работодатель  Кол-во регионов
0                      Яндекс              181
1                  Ростелеком              152
2                  Спецремонт              116
3      Поляков Денис Иванович               88
4                   ООО ЕФИН                71
...                       ...              ...
23170                  СДЕЛКА                0
23171            Alandr Group                0
23172                     СДК                0
23173      Lemon Land Lombard                0
23174                  Махина                0

[23175 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [73]:
# текст запроса
query_5_4 = f'''SELECT       
                     COUNT(*) AS "Кол-во работодателей" /*Подсчитываем кол.-во работодателей, данные вносим в созданый столбец "Кол-во работодателей"*/ 
                 FROM public.employers e /*Из схеме public таблицы "employers", задаём таблице алиас e*/
                 LEFT JOIN public.employers_industries ei ON ei.employer_id = e.id /*Внутреннее присоединение из схемы public таблицы employers_industries,*/
                                                                                  /*задав ей алиас ei, по ключам id и employer_id*/ 
                 WHERE ei.industry_id IS NULL /*Указываем услови при котором работодатель не указал сферу деятельности*/''' 
# Используем f-строки для удобства передачи параметров

In [74]:
# результат запроса
# Используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df_5_4 = pd.read_sql_query(query_5_4, connection)
df_5_4

C:\Users\Professional\AppData\Local\Temp\ipykernel_9196\849233504.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_4 = pd.read_sql_query(query_5_4, connection)


Кол-во работодателей
0                  8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [75]:
# текст запроса
query_5_5 = f'''SELECT
                     e.name AS "Названия компаний", /*Выбираем работодателей по столбецу "name" из таблици e,*/
                                                  /*данные заносим в созданый столбец "Названия компаний"*/      
                     COUNT(ei.industry_id) AS "Кол-во сфер деятельности" /*Подсчитываем кол.-во сфер деятельности из столбцу "industry_id" из таблици ei,*/
                                                  /*данные заносим в созданый столбец "Кол-во сфер деятельности"*/   
                 FROM public.employers e /*Из схеме public таблицы "employers", задаём таблице алиас e*/
                 FULL JOIN public.employers_industries ei ON ei.employer_id = e.id /*Внутреннее присоединение из схемы public таблицы employers_industries,*/
                                                                                  /*задав ей алиас ei, по ключам id и employer_id*/ 
                 GROUP BY e.name /*Группируем по столбцу "Названия компаний"*/
                 HAVING COUNT(ei.industry_id) = 4 /*Устанавливаем значение столбца "Кол-во сфер деятельности" равное 4*/ 
                 ORDER BY 1 /*Сортируем по столбцу Названия компаний"*/
                 OFFSET 2 /*Отсукаем первые две строки*/
                 LIMIT 1 /*Указываем кол.-во выводимых строк*/'''
# Используем f-строки для удобства передачи параметров

In [76]:
# результат запроса
# Используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df_5_5 = pd.read_sql_query(query_5_5, connection)
df_5_5

C:\Users\Professional\AppData\Local\Temp\ipykernel_9196\253154687.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_5 = pd.read_sql_query(query_5_5, connection)


Названия компаний  Кол-во сфер деятельности
0              2ГИС                         4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [78]:
# текст запроса
query_5_6 = f'''SELECT
                     COUNT(ei.employer_id) AS "Кол-во работодателей" /*Подсчитываем кол.-во работодателей из столбцу "employer_id" из таблици ei,*/
                                                                     /*данные заносим в созданый столбец "Кол-во работодателей"*/   
                 FROM public.employers_industries ei /*Из схеме public таблицы "employers_industries", задаём таблице алиас ei*/
                 LEFT JOIN public.industries i ON ei.industry_id = i.id /*Внутреннее присоединение из схемы public таблицы industries,*/
                                                                        /*задав ей алиас i, по ключам id и industry_id*/ 
                 WHERE i.name = 'Разработка программного обеспечения' /*Указываем услови при котором сферой деятельности указано 'Разработка программного обеспечения'*/'''
# Используем f-строки для удобства передачи параметров

In [79]:
# результат запроса
# Используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df_5_6 = pd.read_sql_query(query_5_6, connection)
df_5_6

C:\Users\Professional\AppData\Local\Temp\ipykernel_9196\1807149835.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_6 = pd.read_sql_query(query_5_6, connection)


Кол-во работодателей
0                  3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [80]:
# код для получения списка городов-милионников
import urllib.parse as req #Для получения списка городов милиоников с сайта добовляем еще одну бублиотеку "urllib.parse"

path_common = req.urlparse('https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров') # Выполняем запрос по адрессу страници
# Преоброзование полученой информации и создание списка городов
url = path_common.scheme + '://' + req.quote(path_common.netloc) +  req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)
cities = tuple(pd.read_html(url)[1]['Город'])
cities # Вывод списка городов милиоников

('Москва',
 'Санкт-Петербург',
 'Новосибирск',
 'Екатеринбург',
 'Казань',
 'Нижний Новгород',
 'Красноярск',
 'Челябинск',
 'Самара',
 'Уфа',
 'Ростов-на-Дону',
 'Краснодар',
 'Омск',
 'Воронеж',
 'Пермь',
 'Волгоград')

In [86]:
# текст запроса
query_5_7 = f'''SELECT
                     a.name AS "Регионы-миллионники",/*Выбираем регионы-милионики по столбецу "name" из таблици a,*/
                                                    /*данные заносим в созданый столбец "Регионы-миллионники"*/      
                     COUNT(v.id) "Кол-во вакансий" /*Просчитываем кол.-во вакансий по столбецу "id" из таблици v,*/
                                                  /*данные заносим в созданый столбец "Кол-во вакансий"*/
                 FROM public.vacancies v /*Из схеме public таблицы "vacancies", задаём таблице алиас v*/
                 LEFT JOIN public.employers e on v.employer_id = e.id /*Внутреннее присоединение из схемы public таблицы employers,*/
                                                                        /*задав ей алиас e, по ключам id и employer_id*/
                 LEFT JOIN public.areas a on v.area_id = a.id /*Внутреннее присоединение из схемы public таблицы areas,*/
                                                            /*задав ей алиас a, по ключам id и area_id*/
                 WHERE e.name = 'Яндекс' and a.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань',
                 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск',
                 'Воронеж', 'Пермь', 'Волгоград') /*Указываем услови при котором имя компании указано "Яндекс" а регионы выбираются из полученого списка городов милиоников*/
                 GROUP BY a.name /*Группируем по столбцу "Регионы-миллионники"*/
                 UNION ALL /*Оператор присоединения*/
                 SELECT 
                     'Total', count(*) /*Долбовляем строку "Total" хранящую общее количество вакансий*/
                 FROM public.vacancies v /*Из схеме public таблицы "vacancies", задаём таблице алиас v*/
                 LEFT JOIN public.employers e on v.employer_id = e.id /*Внутреннее присоединение из схемы public таблицы employers,*/
                                                                    /*задав ей алиас e, по ключам id и employer_id*/
                 LEFT JOIN public.areas a on v.area_id = a.id /*Внутреннее присоединение из схемы public таблицы areas,*/
                                                            /*задав ей алиас a, по ключам id и area_id*/
                  WHERE e.name = 'Яндекс' and a.name in ('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань',
                 'Нижний Новгород', 'Красноярск', 'Челябинск', 'Самара', 'Уфа', 'Ростов-на-Дону', 'Краснодар', 'Омск',
                 'Воронеж', 'Пермь', 'Волгоград') /*Указываем услови при котором имя компании указано "Яндекс" а регионы выбираются из полученого списка городов милиоников*/
                  ORDER BY 2 /*Сортируем по кол.-ву вакаесий в порядке возростания*/'''
# Используем f-строки для удобства передачи параметров

In [88]:
# результат запроса
# Используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df_5_7 = pd.read_sql_query(query_5_7, connection)
df_5_7

C:\Users\Professional\AppData\Local\Temp\ipykernel_9196\3779922428.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_7 = pd.read_sql_query(query_5_7, connection)


Регионы-миллионники  Кол-во вакансий
0                 Омск               21
1            Челябинск               22
2           Красноярск               23
3            Волгоград               24
4                Пермь               25
5               Казань               25
6       Ростов-на-Дону               25
7                  Уфа               26
8               Самара               26
9            Краснодар               30
10             Воронеж               32
11         Новосибирск               35
12     Нижний Новгород               36
13        Екатеринбург               39
14     Санкт-Петербург               42
15              Москва               54
16               Total              485

***

# выводы по анализу работодателей
По итогом проведеного анализа можно сказать
 - Компания Яндекс занимает первое место по количеству вакансий. Это указывает на значительное присутствие и активность компании на рынке труда.
 - В регионах, где нет вакансий, большинство работодателей находится в России, что может быть связано с ориентацией данных работодателей на местный рынок труда.
 - Выявлено 8419 работодателей, у которых не указана сфера деятельности. Это может указывать на неполную или некорректную заполненность данных в базе.
 - Компания "Яндекс2 представила 485 вакансий в 16 городах-миллионниках. Это говорит о широком географическом охвате и наличии возможностей для трудоустройства в разных городах.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [91]:
# текст запроса
query_6_1 = f'''SELECT
                     COUNT(v.id) AS "Кол-во вакансий, связанных с DS" /*Просчитываем кол.-во вакансий по столбецу "id" из таблици v,*/
                                                               /*данные заносим в созданый столбец "Кол-во вакансий, связанных с DS"*/
                 FROM public.vacancies v /*Из схеме public таблицы "vacancies", задаём таблице алиас v*/
                 WHERE lower(v.name) LIKE '%данн%' OR lower(v.name) LIKE '%data%' /*Указываем условия поиска названия вакансии, попутно приводя названия вакансий к единному регистру*/''' 
# Используем f-строки для удобства передачи параметров

In [92]:
# результат запроса
# Используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df_6_1 =pd.read_sql_query(query_6_1, connection)
df_6_1

C:\Users\Professional\AppData\Local\Temp\ipykernel_9196\4147534520.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_1 =pd.read_sql_query(query_6_1, connection)


Кол-во вакансий, связанных с DS
0                             1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [93]:
# текст запроса
query_6_2 = f'''SELECT
                     COUNT(v.id) AS "Кол-во вакансий junior DS" /*Просчитываем кол.-во вакансий по столбецу "id" из таблици v,*/
                                                               /*данные заносим в созданый столбец "Кол-во вакансий junior DS"*/
                 FROM public.vacancies v /*Из схеме public таблицы "vacancies", задаём таблице алиас v*/
                 WHERE (lower(v.name) LIKE '%data scientist%' OR lower(v.name) LIKE '%data science%' OR lower(v.name) like '%исследователь данных%' 
                     OR (v.name LIKE '%ML%' AND v.name NOT LIKE '%HTML%') OR lower(v.name) LIKE '%machine learning%' OR lower(v.name) LIKE '%машинн%обучен%')
                     AND (lower(v.name) LIKE '%junior%' OR lower(v.experience) LIKE '%нет опыта%' 
                     OR lower(v.experience) LIKE '%стажировка%') /*Указываем условия поиска названия вакансии и опыта работы, попутно приводя запрашиваемые данные к единному регистру*/'''
# Используем f-строки для удобства передачи параметров

In [94]:
# результат запроса
# Используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df_6_2 =pd.read_sql_query(query_6_2, connection)
df_6_2

C:\Users\Professional\AppData\Local\Temp\ipykernel_9196\3043554268.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_2 =pd.read_sql_query(query_6_2, connection)


Кол-во вакансий junior DS
0                         51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [95]:
# текст запроса
query_6_3 = f'''SELECT
                     COUNT(v.id) AS "Кол-во вакансий DS/SQL или postgres"  /*Просчитываем кол.-во вакансий по столбецу "id" из таблици v,*/
                                                                /*данные заносим в созданый столбец "Кол-во вакансий DS/SQL или postgres"*/
                 FROM public.vacancies v /*Из схеме public таблицы "vacancies", задаём таблице алиас v*/
                 WHERE (lower(v.name) LIKE '%data scientist%' OR lower(v.name) LIKE '%data science%' OR lower(v.name) like '%исследователь данных%' 
                     OR (v.name LIKE '%ML%' AND v.name NOT LIKE '%HTML%') OR lower(v.name) LIKE '%machine learning%' OR lower(v.name) LIKE '%машинн%обучен%')
                     AND (v.key_skills LIKE '%SQL%' OR v.key_skills LIKE '%postgres%') /*Указываем условия поиска названия вакансии и навыкам,*/
                                                                                      /*попутно приводя запрашиваемые данные к единному регистру*/'''
# Используем f-строки для удобства передачи параметров

In [96]:
# результат запроса
# Используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df_6_3 =pd.read_sql_query(query_6_3, connection)
df_6_3

C:\Users\Professional\AppData\Local\Temp\ipykernel_9196\2186801325.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_3 =pd.read_sql_query(query_6_3, connection)


Кол-во вакансий DS/SQL или postgres
0                                  201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [97]:
# текст запроса
query_6_4 = f'''SELECT
                     COUNT(v.id) AS "Кол-во вакансий DS/Python" /*Просчитываем кол.-во вакансий по столбецу "id" из таблици v,*/
                                                                /*данные заносим в созданый столбец "Кол-во вакансий DS/Python"*/
                 FROM public.vacancies v /*Из схеме public таблицы "vacancies", задаём таблице алиас v*/
                 WHERE (lower(v.name) LIKE '%data scientist%' OR lower(v.name) LIKE '%data science%' OR lower(v.name) like '%исследователь данных%' 
                     OR (v.name LIKE '%ML%' AND v.name NOT LIKE '%HTML%') OR lower(v.name) LIKE '%machine learning%' OR lower(v.name) LIKE '%машинн%обучен%')
                     AND v.key_skills LIKE '%Python%' /*Указываем условия поиска названия вакансии и навыки работы с Python,*/
                                                      /*попутно приводя запрашиваемые данные к единному регистру*/'''
# Используем f-строки для удобства передачи параметров

In [98]:
# результат запроса
# Используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df_6_4 =pd.read_sql_query(query_6_4, connection)
df_6_4

C:\Users\Professional\AppData\Local\Temp\ipykernel_9196\1290520132.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_4 =pd.read_sql_query(query_6_4, connection)


Кол-во вакансий DS/Python
0                        351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [99]:
# текст запроса
query_6_5 = f'''SELECT
                     round(AVG(LENGTH(v.key_skills) - LENGTH(REPLACE(v.key_skills, chr(9), '')) + 1), 2) 
                     AS "Среднее кол-во ключевых навыков" /*Просчитываем среднее кол.-во  необходимых навыков в вакансиях DS,*/
                                                        /*данные заносим в созданый столбец "Среднее кол-во ключевых навыков"*/
                 FROM public.vacancies v /*Из схеме public таблицы "vacancies", задаём таблице алиас v*/
                 WHERE (lower(v.name) LIKE '%data scientist%' OR lower(v.name) LIKE '%data science%' OR lower(v.name) like '%исследователь данных%' 
                     OR (v.name LIKE '%ML%' AND v.name NOT LIKE '%HTML%') OR lower(v.name) LIKE '%machine learning%' 
                     OR lower(v.name) LIKE '%машинн%обучен%')/*Указываем условия поиска названия вакансии, попутно приводя запрашиваемые данные к единному регистру*/'''
# Используем f-строки для удобства передачи параметров

In [100]:
# результат запроса
# Используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df_6_5 =pd.read_sql_query(query_6_5, connection)
df_6_5

C:\Users\Professional\AppData\Local\Temp\ipykernel_9196\1519424183.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_5 =pd.read_sql_query(query_6_5, connection)


Среднее кол-во ключевых навыков
0                             6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [101]:
# текст запроса
query_6_6 = f'''SELECT
                     round(AVG(COALESCE((v.salary_to + v.salary_from)/2, v.salary_to, v.salary_from))) 
                     AS "Средняя з/п специалиста DS с опытом от 3 до 6 лет" /*Расчитываем среднюю з/п специалиста DS с опытом от 3 до 6 лет,*/
                                                    /*данные заносим в созданый столбец "Средняя з/п специалиста DS с опытом от 3 до 6 лет"*/
                 FROM public.vacancies v /*Из схеме public таблицы "vacancies", задаём таблице алиас v*/
                 WHERE (lower(v.name) LIKE '%data scientist%' OR lower(v.name) LIKE '%data science%' OR lower(v.name) like '%исследователь данных%' 
                     OR (v.name LIKE '%ML%' AND v.name NOT LIKE '%HTML%') OR lower(v.name) LIKE '%machine learning%' OR lower(v.name) LIKE '%машинн%обучен%')
                     AND (v.salary_to IS NOT NULL OR v.salary_from IS NOT NULL) 
                     AND v.experience = 'От 3 до 6 лет' /*Указываем условия поиска попутно приводя запрашиваемые данные к единному регистру*/'''

# Используем f-строки для удобства передачи параметров

In [102]:
# результат запроса
# Используем метод read_sql_query, чтобы получить результаты запроса сразу в виде датафрейма
df_6_6 =pd.read_sql_query(query_6_6, connection)
df_6_6

C:\Users\Professional\AppData\Local\Temp\ipykernel_9196\3769920193.py:3: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_6 =pd.read_sql_query(query_6_6, connection)


Средняя з/п специалиста DS с опытом 
0                              243115.0

***

# выводы по предметному анализу
Согласно проведеному анализу вакансий DS можно сделать следующие выводы
 - Количество вакансий, завязаных на работы с данными, составляет 1771. Это указывает на наличие спроса на специалистов в data science.
 - Существует 51 подходящая вакансия для начинающего специалиста, что дает возможность для новичков в этой области найти работу и начать свою карьеру.
 - Количество вакансий для DS-специолистов, где SQL или postgre указаны в качестве ключевых навыков, составляет 201. Это указывает на важность знания SQL или postgre для работы в этой области и требования работодателей в этом отношении.
 - Количество вакансий, где Python указывается в качестве ключевого навыка, составляет 351. Это говорит о том, что знание Python востребовано у работодателя и является преимушеством соискателя.
 - Среднее количество ключевых навыков для дата-сайентиста составляет более 6. Это говорит о том, что в этой области требуются специалисты с широким спектром навыков и компетенций.
 - Средняя зарплата для дата-сайентиста с опытом работы от 3 до 6 лет составляет 243115, указывая на то, что опытные специалисты в этой области могут рассчитывать на высокую заработную плату. 

# Общий вывод по проекту

Прежде чем преступить к обобщению полученых результатов проведем еще несколько допольнитедьных анализов имеющихся даных

Д.1 Анализ количества работодателей по сферам деятельности. Выведем первую тройку сфер деятельности по количеству работодателей.

In [110]:
# текст запроса
query_d_1 = f'''SELECT
                    i.name AS "Сфера деятельности", /*Выбираем сферы деятельности по столбецу "name" из таблици i,*/
                                                    /*данные заносим в созданый столбец "Сфера деятельности"*/   
                    COUNT(DISTINCT ei.employer_id) AS "Количество работодателей" /*Подсчитываем кол.-во работодателей по столбцу "employer_id" из таблици ei,*/
                                                                                 /*данные заносим в созданый столбец "Количество работодателей"*/
                 FROM public.employers_industries ei /*Из схеме public таблицы "employers_industries", задаём таблице алиас ei*/
                 JOIN public.industries i ON ei.industry_id = i.id /*Внутреннее присоединение из схемы public таблицы industries,*/
                                                                    /*задав ей алиас i, по ключам id и industry_id*/
                 GROUP BY 1 /*Группируем по столбцу "Сфера деятельности"*/ 
                 ORDER BY 2 DESC /*Сортируем по столбцу "Количество работодателей" в порядке убывания*/
                 LIMIT 3 /*Выводим первые три строки*/'''
# Используем f-строки для удобства передачи параметров
# Для вывода запроса используем метод read_sql_query, чтобы получить результаты сразу в виде датафрейма
df_d_1 =pd.read_sql_query(query_d_1, connection)
df_d_1

C:\Users\Professional\AppData\Local\Temp\ipykernel_9196\250438586.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_d_1 =pd.read_sql_query(query_d_1, connection)


Сфера деятельности  Количество работодателей
0                Разработка программного обеспечения                      3553
1  Системная интеграция,  автоматизации технологи...                      2993
2  Интернет-компания (поисковики, платежные систе...                      1675

Д.2 Проведём анализ количества вакансий по сферам деятельности. Выведем первую тройку сфер деятельности по количеству вакансий.

In [111]:
# текст запроса
query_7_1 = f'''SELECT
                         i.name AS "Сфера деятельности", /*Выбираем сферы деятельности по столбецу "name" из таблици i,*/
                                                         /*данные заносим в созданый столбец "Сфера деятельности"*/   
                         COUNT(v.id) AS "Количество вакансий" /*Подсчитываем кол.-во вакансий по столбцу "id" из таблици v,*/
                                                         /*данные заносим в созданый столбец "Количество вакансий"*/
                 FROM industries i /*Из схеме public таблицы "industries", задаём таблице алиас i*/
                 JOIN employers_industries ei ON i.id = ei.industry_id /*Внутреннее присоединение из схемы public таблицы employers_industries,*/
                                                                    /*задав ей алиас ei, по ключам id и industry_id*/
                 JOIN vacancies v ON ei.employer_id = v.employer_id /*Внутреннее присоединение из схемы public таблицы vacancies,*/
                                                                    /*задав ей алиас v, по ключам employer_id и employer_id*/
                 GROUP BY 1 /*Группируем по столбцу "Сфера деятельности"*/ 
                 ORDER BY 2 DESC /*Сортируем по столбцу "Количество вакансий" в порядке убывания*/
                 LIMIT 3 /*Выводим первые три строки*/'''
# Используем f-строки для удобства передачи параметров
# Для вывода запроса используем метод read_sql_query, чтобы получить результаты сразу в виде датафрейма                 
df =pd.read_sql_query(query_7_1, connection)
df

C:\Users\Professional\AppData\Local\Temp\ipykernel_9196\1436291614.py:17: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df =pd.read_sql_query(query_7_1, connection)


Сфера деятельности  Количество вакансий
0                Разработка программного обеспечения                12499
1  Системная интеграция,  автоматизации технологи...                11034
2  Интернет-компания (поисковики, платежные систе...                 6413

In [112]:
#закрываем соединение после окончания работы
connection.close()

Выводы по дополнительным исследованиям:
- Большое количество работодателей (3553) в сфере "Разработка программного обеспечения" порождает и большое количество вакансий даного напровления (12499), но это также указывает и на больной спрос специолистов в этой области и наличие множества компаний, занимающихся разработкой программного обеспечения.
 - Сдедующей сферой со значительным количеством работодателей (2993) и вакансий (11034) являеться "Системная интеграция, автоматизации технологий", говорит о том, что автоматизация технологических процессов и системная интеграция являются востребованными и перспективными областями.
 - Сфера деятельности "Интернет-компания (поисковики, платежные системы и другие)" имеет меньшее количество вакансий (6413) и работодателей (1675) по сравнению с предыдущими двумя сферами. Однако она все равно представляет собой значительную долю рынка труда и может предоставлять интересные для трудоустройства.

# подведем итог исследования, обобщите выводы

Обобщая полученые данные можно сказать
 - Выборка содержит 49197 вакансий, представлен 23501 работодатель, содержится информация о 1362 регионах по 294 сферам деятельности.
 - Лидерами по количеству вакансий являются мегаполисы, выделяя их значимость на рынке труда и развитость как центров бизнеса и экономики.
 - Среднее значение для нижней зарплатной вилки составляет 71065, а для верхней зарплатной вилки - 110537.
 - Опыт работы от 1 года до 3 лет является наиболее востребованным.
 - Полная занятость является наиболее популярным типом занятости.
 - Регионом с отсудствующими вакансиями, но большим кол.-вом работодателей является Россия, что указывает на ориентацию работодателей данного региона на местный рынок труда.
 - Компания "Яндекс" занимает первое место по количеству вакансий.
 - Полная занятость является наиболее популярным типом занятости.
 - Кол.-во вакансий, связанных с работой с данными, составляет 1771, из которых вакансии с навыками работы в Python составляет 351. Среднее же значение навыков для DS-сайентиста составляет более 6, сам же соискатель с опытом работы от 3 лет может расчитывать в среднем на зарплату в 243115 руб. При этом вакансий для молодых специолистов мало всего 51 вакансия.
 - Сфера деятельности "Разработка программного обеспечения" имеет наибольшее количество вакансий - 12499 и наибольшее количество работодателей - 3553.

 (Все расчеты приведены на момент создания предоставленых таблиц)